### PR TITLE
telemetry: adds the telemetry package and related features.

### PR DESCRIPTION
**Pull request by remeh**
### What does this PR do?

Adds a `telemetry` package in order to add observability metrics on the internal behavior of the Agent.

### Motivation

Having a better observability on the internal of the Agents to constantly ensure the quality and the good behavior of the Agent while working on new releases.

### Additional Notes

The observability metrics are added in #4600

### CODE REVIEW

File: pkg/collector/corechecks/embed/apm.go
Comment (arbll) [inline pkg/collector/corechecks/embed/apm.go]: I wonder if this code is still used since the trace agent lifecycle is managed by systemd

File: pkg/collector/corechecks/embed/jmx/check.go
Comment (arbll) [inline pkg/collector/corechecks/embed/jmx/check.go]: We might want to hardcode this to false. I don't think any metric is "routed" through this check. This thing is just some kind of hack to have a long running jmxfetch.

File: pkg/collector/corechecks/embed/process_agent.go
Comment (arbll) [inline pkg/collector/corechecks/embed/process_agent.go]: Same as apm.go

File: pkg/telemetry/prom_counter.go
Comment (arbll) [inline pkg/telemetry/prom_counter.go]: I don't think this is used ?

File: pkg/telemetry/gauge.go
Comment (arbll) [inline pkg/telemetry/gauge.go]: Note:this might panic if anything violates prometheus naming constraints

**Review by arbll — COMMENTED** [review]
First pass, looks clean and simple 👍

File: pkg/telemetry/prom_counter.go
Comment (remeh) [reply pkg/telemetry/prom_counter.go]: Nice catch!

File: pkg/collector/corechecks/embed/jmx/check.go
Comment (remeh) [reply pkg/collector/corechecks/embed/jmx/check.go]: I don't support it right now, but if later we want to have for instance the execution time of each run of the JMX check, I think we would be able to enable/disable it through this flag because it'll pass by the JMXCheck.GetMetricStats(), WDYT?

File: pkg/collector/corechecks/embed/apm.go
Comment (remeh) [reply pkg/collector/corechecks/embed/apm.go]: That's a good question, I have to look into that.

File: pkg/telemetry/gauge.go
Comment (remeh) [reply pkg/telemetry/gauge.go]: Yes and I think that's the wanted behavior: I want it to fail during the development phase of someone adding new telemetry metrics.

Comment (arbll) [reply pkg/telemetry/gauge.go]: Agreed that it's the good approach as long as none of those parameters are dynamic.

**Comment by albertvaka** [comment]
Should we change `CODEOWNERS` so telemetry is owned by agent core?

**Comment by remeh** [comment]
@albertvaka Absolutely! I've sent a commit.

**Comment by albertvaka** [comment]
I'll approve once someone from agent-core approves.

File: cmd/agent/app/run.go
Comment (clamoriniere) [inline cmd/agent/app/run.go]: I think it can be good to use a local `prometheus.Registry` and then use [`prometheus.HandlerFor()`](https://godoc.org/github.com/prometheus/client_golang/prometheus/promhttp#HandlerFor)(https://godoc.org/github.com/prometheus/client_golang/prometheus#NewRegistry) it will be easer for testing.


> First, in more complex programs, global state is often quite problematic. That's the reason why the metrics constructors in the prometheus package do not interact with the global prometheus.DefaultRegisterer on their own. You are free to use the Register or MustRegister functions to register them with the global prometheus.DefaultRegisterer, but you could as well choose a local Registerer (usually created with prometheus.NewRegistry, but there are other scenarios, e.g. testing).

Comment (remeh) [reply cmd/agent/app/run.go]: I agree that it's a solution but I don't think that it is necessary, especially because we want the `telemetry` pkg to be callable from all parts of the Agent: we would rely on a global (to the package) instance of a `Registry` in the `telemetry` package, thus having the same behavior as using the default one of the prom package. WDYT?

Comment (clamoriniere) [reply cmd/agent/app/run.go]: I agree that it will require a big refactoring about how we initialise the different Agent components: 
* avoid "func init()"
* use dependency injection or add dependency parameters to the different components.

We had the problem in another project when we wanted to validate metrics generation. because it not easy to cleanup metrics from the registry. 
but I think in your case you are in a better position, since you have added a facade in front of the prometheus api so you can add some mocking feature here.

Comment (remeh) [reply cmd/agent/app/run.go]: Yup it'll be relatively easy to change this due to the abstraction, I'm just not sure on what we would win with this change. If it is about the testability:
- if we want to validate metrics generation, we'll just have to create a mock implementation of the telemetry pkg.
- if we want to validate metrics generation of the prometheus package: that's not in the Agent that it must be done, it must be tested in the lib (prometheus) package.
However, I agree that it would not cost much to have a "local" registry in the telemetry package instead of using the one made available by the prometheus lib, I'll send the change.

Comment (remeh) [reply cmd/agent/app/run.go]: Last commit will now use a custom prometheus registry.

File: pkg/collector/corechecks/embed/apm.go
Comment (remeh) [reply pkg/collector/corechecks/embed/apm.go]: This code is still in use on macOS.

File: pkg/collector/corechecks/embed/process_agent.go
Comment (remeh) [reply pkg/collector/corechecks/embed/process_agent.go]: This code is still in use on macOS.

File: pkg/telemetry/gauge.go
Comment (arbll) [inline pkg/telemetry/gauge.go]: nit: maybe we should call the tag parameter `tagsValue` instead to avoid any confusion ?

**Review by arbll — DISMISSED** [review]
Just a nit, should be ready to merge after it 👍

**Review by arbll — DISMISSED** [review]
📦

**Review by clamoriniere — DISMISSED** [review]

**Review by albertvaka — DISMISSED** [review]

**Review by shang-wang — DISMISSED** [review]

**Pull request merged**
